### PR TITLE
Fix bug in validator.fitTest where anyOf schemata were not handled co…

### PR DESCRIPTION
**Pull request by m-glutting**
Fix bug in validator.fitTest where anyOf schemata were not handled correctly

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Is backward-compatible?    | ✔️
| Tests pass?   | ✔️
| Fixed issues  |  -
| Updated README/docs?   | ❌
| Added CHANGELOG entry?   | ✔️

We use json-editor with a schema which roughly looks like this:
```
{
    "definitions": {
        "prop": {
            "anyOf": [
                {
                    "$ref": "#/definitions/A"
                },
                {
                    "$ref": "#/definitions/B"
                }
            ]
        },
        "A": {
            "type": "object",
            "properties": {
                "fooA": {
                    "type": "string"
                },
                "barA": {
                    "type": "string"
                },
                "common": {
                    "type": "string"
                }
            },
            "required": [
                "fooA",
                "barA"
            ]
        },
        "B": {
            "anyOf": [
                {
                    "$ref": "#/definitions/C"
                },
                {
                    "$ref": "#/definitions/D"
                }
            ]
        },
        "C": {
            "type": "object",
            "properties": {
                "fooC": {
                    "type": "string"
                },
                "barC": {
                    "type": "string"
                },
                "common": {
                    "type": "string"
                }
            },
            "required": [
                "fooD",
                "barD"
            ]
        },
        "D": {
            "type": "object",
            "properties": {
                "fooD": {
                    "type": "string"
                },
                "barD": {
                    "type": "string"
                }
            },
            "required": [
                "fooD",
                "barD"
            ]
        }
    }
}

```
We consider the following value to be matched against this schema:
```
{
    "fooC": "test",
    "barC": "test",
    "common": "test"
}
```
The expected behavior would be that `B` is deemed the correct schema for the test value because the test value matches schema `C` which is listed as an `anyOf` option for `B`.

The old implementation of `fitTest` ignores the nested properties of `B` (because `B` itself has no properties). Because we are in an `anyOf` scenario, the `multiple` editor then replaces the previously correctly identified schema (B) by the one with the best `fitTest` result (see https://github.com/json-editor/json-editor/blob/c9a9196d6643e9a05d979093dde0a0332c15bfa6/src/editors/multiple.js#L273). `A` has a better `fitTest` result than `B` because the test value has the `common` property which is included in the properties of `A` but not in the (direct) properties of `B`. The fact that `validator.validate()` failed (i.e. returned a non-empty array of errors) for `A` and the test value is ignored.

The proposed change solves this problem by changing the `fitTest` implementation so that the properties of nested `anyOf` types are also considered. In our example, this leads to a better result for `fitTest` when called with schema `B` compared to schema `A`.

### CODE REVIEW

**Comment by m-glutting** [comment]
@maintainers: What are the next steps? Can someone please review this? Do you need more input from me?

**Comment by schmunk42** [comment]
Sorry, totally missed this. LGTM.

@germanbisurgi WDYT?

**Comment by germanbisurgi** [comment]
I think we need a test and also to consider "oneOf" and "allOf"  into the equation.

**Comment by m-glutting** [comment]
> I think we need a test and also to consider "oneOf" and "allOf" into the equation.

thank you. Yes, I can add a test. I am not sure about the implications on `oneOff` and `allOff`. Can you help me there please?

**Comment by germanbisurgi** [comment]
Actually i think i was wrong about `oneOff` and `allOff` because a fit test make only sense when there are multiple possible schemas that can match but only the fittest should be the one that matches. With `oneOf` only one schema should match and with `allOf` all schemas should match. If you can add a test for it we will be happy to merge the PR :)

**Comment by germanbisurgi** [comment]
@m-glutting i tried to test the your changes by using your example schema (with some modifications) but i did't found differences in behaviour. Probably i am missing some steps. Could you help me to reproduce the issue?

[Here is a direct link to the example](https://json-editor.github.io/json-editor/?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgogiFUJhO3cW1hRsulCCg0jICFMhUpsGhtv2IrsvEEB5IqgBtUAASKSp/CwBiAHoAE3CYeBhYBERogEEQAF8WELCIkBj4okTkuCRogCFsgF0snJBi0pSkV0y0S2SzQkzjbXN2Lh4FOwcnGBcOkjB2zX7CRCgpRIBzbLY8MilZy3mLReX4NYasOjoEVy0dBaXV7IawgEcRGDDYwOIwGYVN7ZA6tjVDomboWapsLy+CJBEChcKEIoJJItNIAYXWeXhUTiSLKqWiABFag10cCugN0X1rhZ9EMbKMdONJqBpqS5tTFLcjusQL82bsOQc7id6Od4Jc9pzDscHlRnq8qO8UAFPmBKbytuiASAiWTTAMiVSBrTDCN3I5nMgpl9deyBkLuQ1fraBfauTK2E8Xm8PtNDRqpES6vU2IhlGAKLxHPYpFaQIkrFIyDxygooKoGIQ8F8oAcyBAACz3IA) and this is the schema i was using:

```json
{
  "type": "object",
  "title": "test",
  "properties": {
    "prop": {
      "anyOf": [
        {
          "$ref": "#/definitions/A"
        },
        {
          "$ref": "#/definitions/B"
        }
      ]
    }
  },
  "definitions": {
    "A": {
      "title": "A",
      "type": "object",
      "properties": {
        "fooA": {
          "type": "string"
        },
        "barA": {
          "type": "string"
        },
        "common": {
          "type": "string"
        }
      },
      "required": [
        "fooA",
        "barA"
      ]
    },
    "B": {
      "title": "B",
      "anyOf": [
        {
          "$ref": "#/definitions/C"
        },
        {
          "$ref": "#/definitions/D"
        }
      ]
    },
    "C": {
      "title": "C",
      "type": "object",
      "properties": {
        "fooC": {
          "type": "string"
        },
        "barC": {
          "type": "string"
        },
        "common": {
          "type": "string"
        }
      },
      "required": [
        "fooC",
        "barC"
      ]
    },
    "D": {
      "title": "D",
      "type": "object",
      "properties": {
        "fooD": {
          "type": "string"
        },
        "barD": {
          "type": "string"
        }
      },
      "required": [
        "fooD",
        "barD"
      ]
    }
  }
}
```

**Comment by m-glutting** [comment]
@germanbisurgi thanks for providing the live example! I think it is well suited to reproduce the problem: If you add properties `fooC` and `barC` instead of `fooA` and `barA` in the "JSON Output" area and then hit "Update Form", I would expect the editor to keep the data the same and accept that `prop` now matches the definition of `B` because it matches the definition of `C`.

Instead, the editor re-adds `fooA` and `barA` and still interprets `prop` to be of type `A`.

You can also use the editor to change the type op Prop from `A` to `B` and then select `C` and remove `fooA` and `barA`. If you hit "Update Form" after these steps, the editor again switches back to type `A` and re-adds `fooA` and `barA`.

I hope that I explained everything correctly, please let me know if I need to be more specific.

**Comment by m-glutting** [comment]
@germanbisurgi I have added a test fixture for the nested any of schema now and I have also adjusted the tests of the validator so that they also check whether the editor changes any valid data (this should never happen). This works fine for all tests except one: the `patternProperties` fixture:
https://github.com/json-editor/json-editor/blob/063d688d3c7635518ba2e0f62bf65eb090448ed4/tests/fixtures/validation.json#L712-L736

The test fails because setting the value `{ testing: 'test' }` does not work correctly. Instead, the editor has an empty object `{ }` as its value.

IMO the test itself (asserting that the editor never changes valid data) is good. We now need to figure out, what is wrong with the case I mentioned above.

What do you think?

**Comment by glutengo** [comment]
I think for this to work, this line should also check whether the property name matches any of the `patternProperties` of the schema.

https://github.com/json-editor/json-editor/blob/063d688d3c7635518ba2e0f62bf65eb090448ed4/src/editors/object.js#L996

**Comment by m-glutting** [comment]
I think for this to work, this line should also check whether the property name matches any of the `patternProperties` of the schema.

https://github.com/json-editor/json-editor/blob/063d688d3c7635518ba2e0f62bf65eb090448ed4/src/editors/object.js#L996

**Comment by m-glutting** [comment]
Pipeline looks good now.

**Comment by germanbisurgi** [comment]
This is looking good so far. I have a last request. Can you add an additional item in the valid section of the test? I want to be sure that also the "D" value is valid also when is not the first schema in the anyOf.

**Review by germanbisurgi — APPROVED** [review]

**Comment by germanbisurgi** [comment]
@schmunk42 I think we can merge this, what do you think?

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 ### 2.5.4
 
 - fixes #997 add accessibility support for string input types
+- fix bug in validator.fitTest where anyOf schemata were not handled correctly
```
Comment (schmunk42) [inline CHANGELOG.md:11]: Just a minor thing, but...

This needs to be moved to unreleased, or?

Comment (m-glutting) [reply CHANGELOG.md]: You are right. I have moved it.

**Review by schmunk42 — APPROVED** [review]

**Comment by schmunk42** [comment]
Thank you very much!

**Pull request merged**